c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -3028,7 +3024,10 @@ class NativeCachingAllocator : public CUDAAllocator {
     allocated_blocks[mutex_shard_id][block->ptr] = block;
   }
 
+  // Variables by memory snapshot
   c10::ApproximateClockToUnixTimeConverter clock_converter;
+  bool record_history = false;
+  TraceEntryRingBuffer annotation_buffer;
```

Comment:
From comments below, it might make sense to have annotations be a different class that trace entries because they share nearly no fields and require extra fields. It should be pretty simple to reproduce or template out the ring buffer nature of this class to allow for that.

Reply:
Sounds good, I've templated out the ring buffer, so we can re-use for a new struct AnnotationEntry.